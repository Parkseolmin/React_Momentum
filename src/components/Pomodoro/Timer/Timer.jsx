import { CircularProgressbar, buildStyles } from 'react-circular-progressbar';
import { CircularProgressbarWithChildren } from 'react-circular-progressbar';
import 'react-circular-progressbar/dist/styles.css';
import { useContext } from 'react';
import { SettingsContext } from 'context/SettingsContext';
import SettingButton from '../SettingButton/SettingButton';
import PlayButton from '../PlayButton/PlayButton';
import PauseButton from '../PauseButton/PauseButton';
import styles from './Timer.module.css';
import { useOptionContext } from 'context/OptionContext';
import Arrow from 'components/Arrow/Arrow';
import { FaArrowRightArrowLeft } from 'react-icons/fa6';
import { useLocation } from 'react-router-dom';
import { useTimer } from 'hooks/useTimer';

const red = 'rgba(88, 187, 211, 0.8)';
const green = 'rgba(34, 187, 140, 0.85)';

export default function Timer() {
  const { workMinutes, breakMinutes, setShowSettings } =
    useContext(SettingsContext);
  const [
    isPaused,
    togglePause,
    mode,
    minutes,
    seconds,
    percentage,
    innerPercentage,
  ] = useTimer(workMinutes, breakMinutes);
  const { showItemBox, setShowItemBox, itemBoxRef, clockRef } =
    useOptionContext();
  const location = useLocation();

  return (
    <div className={styles.progressBox} ref={clockRef}>
      <span
        className={styles.changeArrow}
        onClick={() => setShowItemBox(!showItemBox)}
      >
        <FaArrowRightArrowLeft />
      </span>
      {showItemBox && (
        <Arrow location={location} itemBoxRef={itemBoxRef} styles={styles} />
      )}
      <CircularProgressbarWithChildren
        value={percentage}
        text={`${minutes}:${seconds}`}
        styles={buildStyles({
          textColor: '#fff',
          pathColor: mode === 'work' ? red : green,
          trailColor: 'transparent',
        })}
        className={
          mode === 'work'
            ? `${styles.circularProgressbarWorkMode}`
            : `${styles.circularProgressbarBreakMode}`
        }
      >
        <div
          style={{
            width: '84%',
          }}
        >
          <CircularProgressbar
            value={innerPercentage}
            styles={buildStyles({
              pathColor: mode === 'work' ? red : green,
              trailColor: 'transparent',
            })}
            className={
              mode === 'work'
                ? `${styles.circularProgressbarWorkMode}`
                : `${styles.circularProgressbarBreakMode}`
            }
          />
        </div>
        <div style={{ marginTop: '20px' }} className={styles.btnbox}>
          {isPaused ? (
            <PlayButton onClick={togglePause} />
          ) : (
            <PauseButton onClick={togglePause} />
          )}
        </div>
      </CircularProgressbarWithChildren>

      <div style={{ marginTop: '20px', textAlign: 'center' }}>
        <SettingButton onClick={() => setShowSettings(true)} />
      </div>
    </div>
  );
}
