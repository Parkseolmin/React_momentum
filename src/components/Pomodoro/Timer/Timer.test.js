import { render, screen, fireEvent } from '@testing-library/react';
import { SettingsContext } from 'context/SettingsContext';
import { OptionProvider } from 'context/OptionContext';
import Timer from './Timer';
import { MemoryRouter } from 'react-router-dom';

// Mock the useTimer hook
jest.mock('hooks/useTimer', () => ({
  useTimer: jest.fn(),
}));

describe('Timer Component', () => {
  const mockSettingsContext = {
    workMinutes: 25,
    breakMinutes: 5,
    showSettings: false,
    setShowSettings: jest.fn(),
  };

  const mockUseTimer = require('hooks/useTimer').useTimer;

  beforeEach(() => {
    mockUseTimer.mockReturnValue({
      switchMode: jest.fn(),
      togglePause: jest.fn(),
      isPaused: true,
      mode: 'work',
      minutes: 25,
      seconds: '00',
      percentage: 50,
      innerPercentage: 25,
    });
  });

  it('Snapshot Test', () => {
    const { asFragment } = render(
      <SettingsContext.Provider value={mockSettingsContext}>
        <OptionProvider>
          <MemoryRouter>
            <Timer />
          </MemoryRouter>
        </OptionProvider>
      </SettingsContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders essential components correctly', () => {
    render(
      <SettingsContext.Provider value={mockSettingsContext}>
        <OptionProvider>
          <MemoryRouter>
            <Timer />
          </MemoryRouter>
        </OptionProvider>
      </SettingsContext.Provider>
    );

    // Verify the presence of essential elements
    expect(screen.getByText('Focus')).toBeInTheDocument();
    expect(screen.getByText('Break')).toBeInTheDocument();
    expect(
      screen.getByRole('button', { name: /settings/i })
    ).toBeInTheDocument();
  });

  it('calls switchMode when Focus/Break button is clicked', () => {
    const switchModeMock = jest.fn();
    mockUseTimer.mockReturnValue({
      switchMode: switchModeMock,
      togglePause: jest.fn(),
      isPaused: true,
      mode: 'work',
      minutes: 25,
      seconds: '00',
      percentage: 50,
      innerPercentage: 25,
    });

    render(
      <SettingsContext.Provider value={mockSettingsContext}>
        <OptionProvider>
          <MemoryRouter>
            <Timer />
          </MemoryRouter>
        </OptionProvider>
      </SettingsContext.Provider>
    );

    const focusButton = screen.getByText('Focus');
    const breakButton = screen.getByText('Break');

    fireEvent.click(focusButton);
    expect(switchModeMock).toHaveBeenCalledWith('work');

    fireEvent.click(breakButton);
    expect(switchModeMock).toHaveBeenCalledWith('break');
  });

  it('displays correct progress based on percentage', () => {
    mockUseTimer.mockReturnValue({
      switchMode: jest.fn(),
      togglePause: jest.fn(),
      isPaused: true,
      mode: 'work',
      minutes: 25,
      seconds: '00',
      percentage: 75,
      innerPercentage: 37.5,
    });

    render(
      <SettingsContext.Provider value={mockSettingsContext}>
        <OptionProvider>
          <MemoryRouter>
            <Timer />
          </MemoryRouter>
        </OptionProvider>
      </SettingsContext.Provider>
    );

    const progressBar = screen.getByText(/25:00/i);
    expect(progressBar).toBeInTheDocument(); // Correct time text
  });
});
