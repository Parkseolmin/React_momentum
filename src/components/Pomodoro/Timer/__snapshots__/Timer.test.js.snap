// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer Component Snapshot Test 1`] = `
<DocumentFragment>
  <div
    class="progressBox"
  >
    <span
      class="changeArrow"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 448 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M438.6 150.6c12.5-12.5 12.5-32.8 0-45.3l-96-96c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L338.7 96 32 96C14.3 96 0 110.3 0 128s14.3 32 32 32l306.7 0-41.4 41.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0l96-96zm-333.3 352c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L109.3 416 416 416c17.7 0 32-14.3 32-32s-14.3-32-32-32l-306.7 0 41.4-41.4c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0l-96 96c-12.5 12.5-12.5 32.8 0 45.3l96 96z"
        />
      </svg>
    </span>
    <div
      class="modeChangeBtn"
    >
      <button
        class="focusButton"
      >
        Focus
      </button>
      <button
        class="breakButton"
      >
        Break
      </button>
    </div>
    <div
      data-test-id="CircularProgressbarWithChildren"
    >
      <div
        style="position: relative; width: 100%; height: 100%;"
      >
        <svg
          class="CircularProgressbar circularProgressbarBreakMode"
          data-test-id="CircularProgressbar"
          viewBox="0 0 100 100"
        >
          <path
            class="CircularProgressbar-trail"
            d="
      M 50,50
      m 0,-46
      a 46,46 0 1 1 0,92
      a 46,46 0 1 1 0,-92
    "
            fill-opacity="0"
            stroke-width="8"
            style="stroke: transparent; stroke-dasharray: 289.02652413026095px 289.02652413026095px; stroke-dashoffset: 0px;"
          />
          <path
            class="CircularProgressbar-path"
            d="
      M 50,50
      m 0,-46
      a 46,46 0 1 1 0,92
      a 46,46 0 1 1 0,-92
    "
            fill-opacity="0"
            stroke-width="8"
            style="stroke: rgba(255, 255, 255, 0.3); stroke-dasharray: 289.02652413026095px 289.02652413026095px; stroke-dashoffset: 144.51326206513048px;"
          />
          <text
            class="CircularProgressbar-text"
            style="fill: #fff;"
            x="50"
            y="50"
          >
            25:00
          </text>
        </svg>
        <div
          data-test-id="CircularProgressbarWithChildren__children"
          style="position: absolute; width: 100%; height: 100%; margin-top: -100%; display: flex; flex-direction: column; justify-content: center; align-items: center;"
        >
          <div
            style="width: 90%;"
          >
            <svg
              class="CircularProgressbar circularProgressbarWorkMode"
              data-test-id="CircularProgressbar"
              viewBox="0 0 100 100"
            >
              <path
                class="CircularProgressbar-trail"
                d="
      M 50,50
      m 0,-46
      a 46,46 0 1 1 0,92
      a 46,46 0 1 1 0,-92
    "
                fill-opacity="0"
                stroke-width="8"
                style="stroke: transparent; stroke-dasharray: 289.02652413026095px 289.02652413026095px; stroke-dashoffset: 0px;"
              />
              <path
                class="CircularProgressbar-path"
                d="
      M 50,50
      m 0,-46
      a 46,46 0 1 1 0,92
      a 46,46 0 1 1 0,-92
    "
                fill-opacity="0"
                stroke-width="8"
                style="stroke: rgba(255, 255, 255, 0.3); stroke-dasharray: 289.02652413026095px 289.02652413026095px; stroke-dashoffset: 216.76989309769573px;"
              />
            </svg>
          </div>
          <div
            class="btnbox"
          >
            <button
              class="playbtn"
            >
              <p>
                START
              </p>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="settingBtn"
    >
      <button
        class="withText"
      >
        <svg
          class="w-6 h-6"
          fill="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M11.078 2.25c-.917 0-1.699.663-1.85 1.567L9.05 4.889c-.02.12-.115.26-.297.348a7.493 7.493 0 0 0-.986.57c-.166.115-.334.126-.45.083L6.3 5.508a1.875 1.875 0 0 0-2.282.819l-.922 1.597a1.875 1.875 0 0 0 .432 2.385l.84.692c.095.078.17.229.154.43a7.598 7.598 0 0 0 0 1.139c.015.2-.059.352-.153.43l-.841.692a1.875 1.875 0 0 0-.432 2.385l.922 1.597a1.875 1.875 0 0 0 2.282.818l1.019-.382c.115-.043.283-.031.45.082.312.214.641.405.985.57.182.088.277.228.297.35l.178 1.071c.151.904.933 1.567 1.85 1.567h1.844c.916 0 1.699-.663 1.85-1.567l.178-1.072c.02-.12.114-.26.297-.349.344-.165.673-.356.985-.57.167-.114.335-.125.45-.082l1.02.382a1.875 1.875 0 0 0 2.28-.819l.923-1.597a1.875 1.875 0 0 0-.432-2.385l-.84-.692c-.095-.078-.17-.229-.154-.43a7.614 7.614 0 0 0 0-1.139c-.016-.2.059-.352.153-.43l.84-.692c.708-.582.891-1.59.433-2.385l-.922-1.597a1.875 1.875 0 0 0-2.282-.818l-1.02.382c-.114.043-.282.031-.449-.083a7.49 7.49 0 0 0-.985-.57c-.183-.087-.277-.227-.297-.348l-.179-1.072a1.875 1.875 0 0 0-1.85-1.567h-1.843ZM12 15.75a3.75 3.75 0 1 0 0-7.5 3.75 3.75 0 0 0 0 7.5Z"
            fill-rule="evenodd"
          />
        </svg>
        Settings
      </button>
    </div>
  </div>
</DocumentFragment>
`;
